src/pydrex/tensors.py

```diff
@@ -0,0 +1,106 @@
+- `voigt_matrix_to_vector`
+- `rotate`
+
+"""
+__all__ = [
+    "PERMUTATION_SYMBOL",
+    "upper_tri_to_symmetric",
+    "voigt_to_elastic_tensor",
+    "elastic_tensor_to_voigt",
+    "voigt_matrix_to_vector",
+    "rotate",
+]
+
+import numpy as np
+import numba as nb
+
+
+PERMUTATION_SYMBOL = np.array(
+    [
+        [[0.0, 0.0, 0.0], [0.0, 0.0, 1.0], [0.0, -1.0, 0.0]],
+        [[0.0, 0.0, -1.0], [0.0, 0.0, 0.0], [1.0, 0.0, 0.0]],
+        [[0.0, 1.0, 0.0], [-1.0, 0.0, 0.0], [0.0, 0.0, 0.0]],
+    ]
+)
+
+
+def upper_tri_to_symmetric(arr):
+    """Create symmetric array using upper triangle of input array."""
+    # <https://stackoverflow.com/questions/58718365/fast-way-to-convert-upper-triangular-matrix-into-symmetric-matrix>
+    upper_tri = np.triu(arr)
+    return np.where(upper_tri, upper_tri, upper_tri.transpose())
+
+
+def voigt_to_elastic_tensor(matrix):
+    """Create 4-th order elastic tensor from an equivalent voigt matrix."""
+    tensor = np.empty((3, 3, 3, 3))
+    for p in range(3):
+        for q in range(3):
+            delta_pq = 1 if p == q else 0
+            i = (p + 1) * delta_pq + (1 - delta_pq) * (7 - p - q) - 1
+            for r in range(3):
+                for s in range(3):
+                    delta_rs = 1 if r == s else 0
+                    j = (r + 1) * delta_rs + (1 - delta_rs) * (7 - r - s) - 1
+                    tensor[p, q, r, s] = matrix[i, j]
+    return tensor
+
+
+def elastic_tensor_to_voigt(tensor):
+    """Create a 6x6 voigt matrix from an equivalent 4-th order elastic tensor."""
+    matrix = np.zeros((6, 6))
+    matrix_indices = np.zeros((6, 6))
+    for p in range(3):
+        for q in range(3):
+            delta_pq = 1 if p == q else 0
+            i = (p + 1) * delta_pq + (1 - delta_pq) * (7 - p - q) - 1
+            for r in range(3):
+                for s in range(3):
+                    delta_rs = 1 if r == s else 0
+                    j = (r + 1) * delta_rs + (1 - delta_rs) * (7 - r - s) - 1
+                    matrix[i, j] += tensor[p, q, r, s]
+                    matrix_indices[i, j] += 1
+    matrix /= matrix_indices
+    return (matrix + matrix.transpose()) / 2
+
+
+def voigt_matrix_to_vector(matrix):
+    """Create the 21-component Voigt vector equivalent to the 6x6 Voigt matrix."""
+    vector = np.zeros(21)
+    for i in range(3):
+        vector[i] = matrix[i, i]
+        vector[i + 3] = np.sqrt(2) * matrix[(i + 1) % 3, (i + 2) % 3]
+        vector[i + 6] = 2 * matrix[i + 3, i + 3]
+        vector[i + 9] = 2 * matrix[i, i + 3]
+        vector[i + 12] = 2 * matrix[(i + 2) % 3, i + 3]
+        vector[i + 15] = 2 * matrix[(i + 1) % 3, i + 3]
+        vector[i + 18] = 2 * np.sqrt(2) * matrix[(i + 1) % 3 + 3, (i + 2) % 3 + 3]
+    return vector
+
+
+@nb.jit(nopython=True)
+def rotate(tensor, rotation):
+    rotated_tensor = np.zeros((3, 3, 3, 3))
+    for i in range(3):
+        for j in range(3):
+            for k in range(3):
+                for l in range(3):
```

Comment:
Using `l` as a variable is ambiguous. I let you judge.

Reply:
You mean with capital `I`? OK I can change it back, wasn't sure.

Reply:
https://www.flake8rules.com/rules/E741.html